src/System.Globalization.Calendars/tests/CalendarHelpers.cs

```diff
@@ -2,88 +2,209 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using Xunit;
 
 namespace System.Globalization.Tests
 {
     public static class CalendarHelpers
     {
-        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
+        public static Calendar[] s_calendars = new Calendar[]
+        {
+            new ThaiBuddhistCalendar(),
+            new JapaneseCalendar()
+        };
+
         private static int MinEra(Calendar calendar) => calendar.GetEra(calendar.MinSupportedDateTime);
+        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
 
-        private static int YearOffset(Calendar calendar) => calendar.GetYear(calendar.MinSupportedDateTime) - 1;
+        private static int MaxRelativeYearInEra(Calendar calendar, int era)
+        {
+            int[] eras = calendar.Eras;
+            if (eras.Length == 1 || era == eras[0])
+            {
+                return MaxDateTimeInEra(calendar, era);
+            }
+            int difference = MaxDateTimeInEra(calendar, era) - MinDateTimeInEra(calendar, era) + 1;
+            if (era != MaxEra(calendar))
+            {
+                difference += 1;
+            }
+            return difference;
+        }
 
-        public static void AddYears_Invalid(Calendar calendar)
+        private static int MinDateTimeInEra(Calendar cal, int era)
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MaxSupportedDateTime, 1));
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MinSupportedDateTime, -1));
+            int[] eras = cal.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1)
+            {
+                return cal.GetYear(cal.MinSupportedDateTime);
```

Comment:
we need to have this written regardless of the calendar we are using. we don't have to restrict it to this set of calendars. just make sure either it return calendar year or Gregorian year and then look at al usage and make sure it is used as intended.